### PR TITLE
Add Local-Ic Relayer Edge Case Tests for Remote Chain Splitter

### LINKED ISSUE #235: remote-chain splitter local interchaintest
blocked by #230 

currently our remote-chain splitter testing is done in two ways:

- explicitly with unit-tests
- implicitly with our current interchaintests

with local-interchain in place, we can do better.

this should involve, but not be limited to, testing various edge cases related to relayers.

we should test timeouts, relayers going down, and anything that can happen as a result of the sudo entrypoint in the contract.

### PR DESCRIPTION
**Pull request by dowlandaiello**
Closes #235 . This PR implements relayer edge case tests in local-interchain for the remote chain splitter. These tests include:

- Happy path test: ensure the remote chain splitter creates an ICA when the relayer is up
- Timeout handling test: ensure the remote chain splitter does not advance when the relayer is down (causing a packet timeout)
- Recovery test: ensure the remote chain splitter can continue after a timed out packet to create the ICA when the relayer is restored

## Failing Tests

Some local-ic tests are not currently working (namely the two party osmo pol test). However, all of the remote chain splitter tests are working. Fixing the two party osmo pol test is not within the scope of this PR, so I did not fix them.

## Overview on Changes / Additions

* `local-interchaintest/Cargo.toml` - Added the remote chain splitter as a dependency to local-ic so that it can be tested
* `Cargo.lock` - See above about `Cargo.toml`
* `local-interchaintest/src/main.rs` - Added the remote chain splitter tests as tests run by local-ictest-e2e
* `local-interchaintest/src/tests/mod.rs`, `local-interchaintest/src/tests/remote_chain_splitter/mod.rs` - Self explanatory
* `local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs` - See above on the 3 tests I added. *Of note:* two functions in this file, `start_relayer` and `stop_relayer` may be useful in other local-ic relayer tests, so moving them to a utils file could be prudent to prevent test dependencies across tests.

### Macro Usage

There is a very short macro in `remote_chain_splitter.rs`. I can remove it if necessary. The macro just consolidates some redundant timeout checking logic in a way that allows breaking from the timeout check on successful ICA creation, which would be slightly more convoluted with a function. IMO, it's short and relatively useful, but not necessary, so it can be removed if desired.

### CODE REVIEW

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+use serde_json::Value;
+use std::{
+    collections::BTreeMap,
+    time::{Duration, Instant},
+};
+use valence_covenant_single_party_pol::msg::DEFAULT_TIMEOUT;
+
+use log::info;
+
+struct ContractInfo {
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:21]: Why you need this? I think you can reuse the Contract struct in localic-std. For the execute of the Cosmwasm object that you are creating later you don't need the code_id so it would be fine

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+    uploader
+        .send_with_local_cache(VALENCE_PATH, NEUTRON_CHAIN_NAME, LOCAL_CODE_ID_CACHE_PATH)
+        .unwrap();
+
+    uploader
+        .send_with_local_cache(ASTROPORT_PATH, NEUTRON_CHAIN_NAME, LOCAL_CODE_ID_CACHE_PATH)
+        .unwrap();
+
+    uploader.send().unwrap();
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:103]: I think this line is not needed and can be removed

**Review by keyleu — CHANGES_REQUESTED** [review]

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+fn make_remote_chain_splitter(test_ctx: &mut TestContext) -> Result<ContractInfo, LocalError> {
+    let atom_denom = test_ctx.get_native_denom().src(GAIA_CHAIN_NAME).get();
+    let uatom_contribution_amount: u128 = 5_000_000_000;
+
+    let split_config: Vec<(String, SplitConfig)> = vec![(
+        atom_denom.clone(),
+        SplitConfig {
+            receivers: BTreeMap::from_iter(vec![
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:115]: nit: You can rewrite this as 
BTreeMap::from([
            (ACC1_ADDRESS_NEUTRON.to_owned(), Decimal::percent(50)),
                (ACC2_ADDRESS_NEUTRON.to_owned(), Decimal::percent(50)),
            ])
  maybe it's more readable.

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    })
+}
+
+fn fund_remote_chain_splitter(
+    test_ctx: &mut TestContext,
+    remote_chain_splitter: &ContractInfo,
+) -> Result<(), LocalError> {
+    let neutron = test_ctx.get_chain(NEUTRON_CHAIN_NAME);
+
+    neutron
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:168]: localic-std has a send helper where you can do this so you don't need to create it manually:
example:
    send(
        test_ctx
            .get_request_builder()
            .get_request_builder(GAIA_CHAIN_NAME),
        DEFAULT_KEY,
        &party_a_deposit_address,
        &[Coin {
            denom: atom_denom.clone(),
            amount: Uint128::new(uatom_contribution_amount),
        }],
        &Coin {
            denom: atom_denom.clone(),
            amount: Uint128::new(5000),
        },
    )
    .unwrap();

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+            })
+            .unwrap()
+            .as_str(),
+            "valence_remote_chain_splitter",
+            None,
+            "",
+        )
+        .unwrap();
+
+    Ok(ContractInfo {
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:156]: In line with the first comment, here you can just return the Contract object you get from instantiate because you don't need the code_id for your tests.

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+) -> Result<(), LocalError> {
+    // Stop the relayer
+    stop_relayer(test_ctx);
+
+    let neutron = test_ctx.get_chain(NEUTRON_CHAIN_NAME);
+
+    let remote_chain_splitter_contract = CosmWasm::new_from_existing(
+        &neutron.rb,
+        None,
+        Some(remote_chain_splitter.code_id),
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:196]: In line with the previous one, you can pass None here because it's only used for instantiating, not for executing

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+    let neutron = test_ctx.get_chain(NEUTRON_CHAIN_NAME);
+
+    // Advance the splitter.
+    // This should trigger SudoMsg::OpenAck, which will set the ContractState to IcaCreated
+
+    // The state should be IcaCreated
+    let remote_chain_splitter_contract = CosmWasm::new_from_existing(
+        &neutron.rb,
+        None,
+        Some(remote_chain_splitter.code_id),
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:266]: Same as above

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+
+    Ok(())
+}
+
+fn start_relayer(test_ctx: &mut TestContext) {
+    let neutron = test_ctx.get_chain(NEUTRON_CHAIN_NAME);
+
+    reqwest::blocking::Client::default()
+        .post(&neutron.rb.api)
+        .json(&serde_json::json!({ "chain_id": NEUTRON_CHAIN_ID, "action": "start-relayer"}))
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:77]: good to know !

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+        .send()
+        .unwrap();
+}
+
+fn stop_relayer(test_ctx: &mut TestContext) {
+    let neutron = test_ctx.get_chain(NEUTRON_CHAIN_NAME);
+
+    reqwest::blocking::Client::default()
+        .post(&neutron.rb.api)
+        .json(&serde_json::json!({ "chain_id": NEUTRON_CHAIN_ID, "action": "stop-relayer"}))
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:87]: Nice!

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs (hunk 8 above)
Comment (keyleu) [reply local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:77]: this could maybe be moved to the locallic-utils repo we have btw, wdyt?

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+fn upload_contracts(test_ctx: &mut TestContext) -> Result<(), LocalError> {
+    let mut uploader = test_ctx.build_tx_upload_contracts();
+
+    uploader
+        .send_with_local_cache(VALENCE_PATH, NEUTRON_CHAIN_NAME, LOCAL_CODE_ID_CACHE_PATH)
+        .unwrap();
+
+    uploader
```
Comment (keyleu) [inline local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:99]: I don't think you should upload all valence and astroport contracts here since you are only using the remote splitter. When running all the tests together it won't matter because we are using a trick to cache all the code ids but if you run this test alone without the rest you will upload all of them and it will take a long time. I'd just upload the remote splitter and that's it, but up to you

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs (hunk 2 above)
Comment (dowlandaiello) [reply local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:103]: :+1:

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs (hunk 8 above)
Comment (dowlandaiello) [reply local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:77]: I think this is a good idea! I will make a PR today.

File: local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+
+struct ContractInfo {
+    code_id: u64,
+    contract_addr: String,
+}
+
+/// How long we will wait until the ICA must have been created, or timed out
+const ICA_TIMEOUT: Duration = Duration::from_secs(30);
+
+macro_rules! with_poll_timeout {
```
[the remark this answers is not in the document]
Comment (dowlandaiello) [reply local-interchaintest/src/tests/remote_chain_splitter/remote_chain_splitter.rs:29]: Helper function would be longer + the macro is pretty simple + the macro allows breaking from the loop manually. IMO, what's the point of having macros in your language if you can never use them?

**Review by keyleu — APPROVED** [review]
LGTM
